core/services/ocr2/plugins/ocr2keeper/evm21/upkeepstate/store.go

```diff
@@ -200,13 +212,27 @@ func (u *upkeepStateStore) upsertStateRecord(ctx context.Context, workID string,
 
 	u.cache[workID] = record
 
-	return u.orm.InsertUpkeepState(persistedStateRecord{
+	u.pendingRecords = append(u.pendingRecords, persistedStateRecord{
 		UpkeepID:            utils.NewBig(upkeepID),
 		WorkID:              record.workID,
 		CompletionState:     uint8(record.state),
 		IneligibilityReason: reason,
 		InsertedAt:          record.addedAt,
-	}, pg.WithParentCtx(ctx))
+	})
+
+	if len(u.pendingRecords) == flushSize {
+		return u.flush(ctx)
```

Comment:
Sure! So would we want to run the flush in the background, and then continue to collecting more records, but block the next flush until the previous flush completes?

Reply:
not sure fully, will need to consider what happens during errors

Reply:
It looks like when we call the post processors in the plugin, we only log out the errors, we don't act on them further. So maybe, if we encounter an error when we flush, we could send the error down a channel and have the `Start` function read from the channel, and log any errors encountered? So that we still log the error, but just in a different place? Though it probably won't be clear that a post processor triggered the error?

Reply:
yeah potentially. We also need to consider if we retry flushing it or not, o/w the queue size can keep on growing

Reply:
maybe the simplest for now is to keep it as is. 
One thing i'm not sure if we need to do is change to `if len(u.pendingRecords) >= flushSize {` ?